### ISSUE TITLE
Provide a custom Asciidoctor macro that reduces the boilerplate required to include snippets

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-restdocs#323]
The basic idea is that rather than using an `include` for each individual snippet, you can use a single macro to include all of the snippets for a particular operation that's been documented. For added flexibility, the macro could use attributes to control which snippets are included.

I prototyped this a while ago as a custom target for the `include` macro: https://github.com/wilkinsona/springrestdocs/commit/6657af0d590be2673f05d7ecd8f1a91244751edc. I think this could be further refined so that it becomes something like `operation::{snippets}/operation-name`.

**Comment by capncodewash** [comment]
This would be very useful.

**Comment by meistermeier** [comment]
@wilkinsona can you take a look at the link? I get a 404. If you don't mind I could take care of this (if we exclude the "automagical" path detection ;) )

**Comment by meistermeier** [comment]
Now I finally got some time to think about this issue and playing around with some possible solutions. I don't think that a "one-include-includes-all" approach fits the rest-docs use-case but introducing categories / operations groups (no better naming found) additional to explicit named snippets might be a useful option.

    include::restdocs:index-example[category='response']

would be the same as (just as an example)

    include::restdocs:index-example[snippets='response-fields, curl-request, http-response']

What is your opinion?

**Comment by wilkinsona** [comment]
I think it could be a useful option, but I'd prefer to tackle it separately and use the `snippets` attribute as a starting point.

**Comment by meistermeier** [comment]
In contrast to the initial idea to use a block macro processor to provide the possibility to use `operation::{snippets}/operation-name` I kept using the include processor. The problem with converting raw input manually and missing ouput/backend format information mentioned in [gh-297 ](https://github.com/spring-projects/spring-restdocs/issues/297#issuecomment-256116925) still exists in asciidoctorj.

**Comment by wilkinsona** [comment]
Ideally, I'd like to implement this as a block macro. I spoke to @mojavelinux about it a few months ago and Dan thought it should be possible but I never figured out how to do it. I thought I'd realised it was possible using the Ruby-based API but while trying to refresh my memory today I think that may have just been wishful thinking.

@mojavelinux do you have any cycles available to help us out here? In a nutshell, I'd like to be able to provide a block macro (or anything else that's similarly concise) that produces [these lines](https://github.com/spring-projects/spring-restdocs/blame/master/samples/rest-notes-spring-hateoas/src/docs/asciidoc/api-guide.adoc#L102-L120) automatically.

**Comment by meistermeier** [comment]
May you have a look at this diff: https://github.com/spring-projects/spring-restdocs/compare/master...meistermeier:block_macro

I can now use `srd::index-example[snippets=response-fields,http-response]` (https://github.com/spring-projects/spring-restdocs/compare/master...meistermeier:block_macro#diff-fa264ba69ca4e20ed2df763d8cc57074R13) to include the both defined snippets.

Creating a new Asciidoctor instance and remove all registered extensions for converting the snippets helped a lot. (https://github.com/spring-projects/spring-restdocs/compare/master...meistermeier:block_macro#diff-c21f76d1d4ce6afa2781db5b77fc3f31R46)

As you see this is just a quick solution (some println etc.) and needs some more code love and level attribute feature. If you think that this is a possible way to go, I will create a new or update the open pull request.

**Comment by wilkinsona** [comment]
> Creating a new Asciidoctor instance and remove all registered extensions for converting the snippets helped a lot. 

Intriguing. What does removing all the registered extensions do? Does it address the problem of potentially getting HTML output in a PDF?

**Comment by meistermeier** [comment]
Unfortunately I currently cannot determine if this problem still exists. The new instance creation with unregistering all extensions solved a problem related to the directory resolver (every include render failed because it could not resolve the projectdir).
Another strange problem I noticed was that the asciidoctor instance could not be created just once for the instance but has to be done in the `process` method. Tests or usage of the processor just hang somewhere and the cpu utilisation ramps up to infinity. ;)
Right now I am struggling with some warnings and somehow failing imports from asciidoctor. After that I hope I can give a detailed comment.

**Comment by wilkinsona** [comment]
Thanks for the details. That sounds like that approach is swimming against the tide quite a bit. I'd prefer to figure out another way to do this or not implement it at all.

**Comment by mojavelinux** [comment]
> The basic idea is that rather than using an include for each individual snippet, you can use a single macro to include all of the snippets for a particular operation that's been documented.

Exactly consistent with my vision for the extension system, specifically the block macro. :+1:

**Comment by mojavelinux** [comment]
Btw, a block macro works when you are including a structural whole. In this case, you are. You are adding several sections. The include is appropriate when you aren't sure what you are including and how it will fit with the rest of the document. That's because an include is a preprocessor, whereas a block macro is an AST contributor (it contributes nodes to the AST as it is being constructed).

In other words, a block macro must be used when what is being included is one or more whole blocks, being inserted somewhere in the AST.

**Comment by mojavelinux** [comment]
If you're going to write an extension for Asciidoctor in Java, I strongly urge you to use AsciidoctorJ 1.6 alpha. The extension API and AST in AsciidoctorJ was just a prototype. We learned a ton from it, then Robert completely overhauled it in 1.6 based on what we learned. I'd go so far as to say I actually think it's risky to write an extension in Java against AsciidoctorJ 1.5.4 at this point.

If you want to avoid that drama for now (which is understandable), you can write the extension in Ruby, which you can still load into AsciidoctorJ. Writing an extension in Ruby does not limit your ability to use it in AsciidoctorJ or with the Gradle or Maven environment. It's really just a language / API preference. In fact, you can even import Java classes into the Ruby script, so you really don't lose any ability at all. And you can always port the code to Java later (once AsciidoctorJ 1.6 really stabilizes). Just giving you options.

**Comment by meistermeier** [comment]
Thanks @mojavelinux for the input. Why not leaving the comfort zone and doing some more Ruby 😀 ? Robert told me at the end of October that there is no real release plan for AsciidotorJ 1.6 and if this is still current a "native" extension might be the solution for the problem.
One question is still remains open (for me): If I do a render the standard result (string) is HTML. Will this block be converted by the backend-modules later if the backend is PDF etc. or do we have to respect this within the extension?

**Comment by mojavelinux** [comment]
While the current approach to parse and convert the snippets file in the extension and insert it as a passthrough block works, it's really not a tenable approach. I've seen this done before and it it's quite inefficient. The extension gives you access to the current parser. The right approach is to read the file and feed it to the parser. It will then incorporate the new blocks into the tree it is building.

You can find a very simple example of this in the extensions lab (imagine the block reader is replaced by a file reader):

* https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/master/lib/compound-block.rb

Though absent from AsciidoctorJ 1.5.4, AsciidoctorJ 1.6 gives you the ability to use the current parser from an extension. I have a more complete example in Java.

```java
public Object process(AbstractBlock parent, Reader reader, Map<String, Object> attrs) {
    Map<String, Object> inheritedAttrs = new HashMap<>();
    Map<Object, Object> options = new HashMap<>();
    String initialContent = null;
    AbstractBlock container = createBlock(parent, "open", initialContent, inheritedAttrs, options);
    parseContent(container, reader.readLines());
    return container;
}
```

(Obviously, it's a lot simpler in Ruby. We're still working on the Java API).

**Comment by mojavelinux** [comment]
> Robert told me at the end of October that there is no real release plan for AsciidotorJ 1.6 

That's something we need to get serious about. I've been secure more resources so we can move this along faster.

**Comment by mojavelinux** [comment]
>  If I do a render the standard result (string) is HTML. 

At all costs, you should avoid converting anything during the parsing phase (which is when the block macro runs). You want to build pure AST. That's got to be the goal. You want to be able to hand off a pure AST to the converter so it has the option to convert to any format it knows how to.

**Comment by mojavelinux** [comment]
In other words, the pass block is a brute force, everything else failed, just do it, sort of solution.

**Comment by mojavelinux** [comment]
Note that after the parseContent call, you have an AST fragment. There is no obligation to return that fragment. You can pick that fragment apart and then copy parts of it to another node. parseContent is just a nice helper to unleash the parser on a source string and get an AST out of it so you can further process it.

(In the future, I think we need to expose a parser utility that provides more control than parseContent does. That's just a something rather than nothing API atm).

**Comment by meistermeier** [comment]
> At all costs, you should avoid converting anything during the parsing phase

> In other words, the pass block is a brute force, everything else failed, just do it, sort of solution.

Since there is no real access for the use case the issue addresses this was my only working solution to come up with. But hey..now is Ruby time.
Thanks for the input.

**Comment by mojavelinux** [comment]
:+1:

**Comment by mojavelinux** [comment]
Once you have the Ruby script, you load it into AsciidoctorJ in the same way the Gradle plugin does it.

https://github.com/asciidoctor/asciidoctor-gradle-plugin/blob/cecab658d6f64d14bbbd7921f31acc817be2d477/src/main/groovy/org/asciidoctor/gradle/AsciidoctorTask.groovy#L585-L587

That method takes an absolute or relative path or the name of a gem on the load path. See http://static.javadoc.io/org.asciidoctor/asciidoctorj/1.6.0-alpha.3/org/asciidoctor/Asciidoctor.html#requireLibrary-java.lang.String...-

**Comment by mojavelinux** [comment]
The way I tend to work is that I write the extension in Ruby and if I decide I want to move it to Java, then I rewrite it. It's not just a language preference. The reason is, the Ruby API is still more complete and more concise, so I can more easily see what I am doing. I want the Java API to feel that way, but we can't lie to ourselves yet and say it is ;)

**Comment by mojavelinux** [comment]
Let me know if you get to a point where you need help and I'll jump back in. To make it go faster, point me to a sample project that I can use to play with the code.

**Comment by wilkinsona** [comment]
Thank you very much, @mojavelinux.

I'm comfortable with using Ruby and have something prototyped that embeds a Ruby script in a jar and loads it into Asciidoctor. When I said above that "I thought I'd realised it was possible using the Ruby-based API but while trying to refresh my memory today I think that may have just been wishful thinking", it's `parse_content` that I'd discovered and then forgotten about. Thank you for the reminder.

**Comment by wilkinsona** [comment]
> Let me know if you get to a point where you need help and I'll jump back in. To make it go faster, point me to a sample project that I can use to play with the code.

Thanks for the offer, @mojavelinux. Unfortunately, it hasn't taken me long to get to the point where I could so with some help please. I have a custom block macro that's adding content to its parent, but I can't get it to recognize the the content should be a section with some content. Instead, I get two paragraphs.

In the interests of simplicity, I've stripped things right back to a pure Ruby implementation in a fork of the extensions lab. The code for what I've tried thus far is in [this branch](https://github.com/wilkinsona/asciidoctor-extensions-lab/tree/restdocs-operation).

Running it produces this output:

```
$ asciidoctor -r ./lib/restdocs-operation.rb lib/restdocs-operation/sample.adoc
#<Asciidoctor::Block@70126376650320 {context: :paragraph, content_model: :simple, style: nil, lines: 1}>
#<Asciidoctor::Block@70126376670980 {context: :paragraph, content_model: :simple, style: nil, lines: 1}>
```

What do I need to do to get `=== Subsection` to be recognised as a section?

**Comment by mojavelinux** [comment]
I'll have a look asap. I think I know what you need, but I want to fiddle
so I give you the right pointers. Stay tuned.

**Comment by mojavelinux** [comment]
Aha. The problem is that right now, the `parse_content` method assumes it's only dealing with non-section content. Upon reflection, this feels like a bad assumption. But we'll have to deal with it upstream.

If what you are parsing includes sections, you have 3 options:

1. Use lower-level APIs to parse both sections and block content
2. Parse the content as a standalone document, then import the nodes from that AST
3. Use discrete section titles (aka headings)

I don't think you really want (3), and (1) is way too risky. So let's focus on (2).

```ruby
process do |parent, reader, attrs|
  content = %q(before

=== Subsection

Some more content)
  doc = parent.document
  fragment = Asciidoctor.load content,
      safe: doc.options[:safe],
      attributes: { 'fragment' => '' }
  fragment.blocks.each do |b|
    b.parent = parent
    parent << b
  end
  nil
end
```

The fragment attribute is important because it informs the parser that the source is a document fragment and should not be validated as a standalone document. Specifically, this allows the top-level section titles to start at any level.

It's clear we need a method in the extensions API that performs a similar operation. That method would also be careful to carry over all processor options so that the parser is operating in the same configuration. But this code is sufficient portable for your needs, I think.

**Comment by wilkinsona** [comment]
Thanks very much, @mojavelinux. That works beautifully

**Comment by wilkinsona** [comment]
@meistermeier Do you still have the time and enthusiasm to take another run at this? Totally understand if you don't as you've already expended quite a bit of energy on it.

**Comment by meistermeier** [comment]
No problem. I will give it a try tomorrow (or tonight) and give you feedback.

**Comment by meistermeier** [comment]
I have tested the approach above but it won't work with PDFs (the only other backend I tested). There is nothing in the logs or document where the macro gets called.
With the input from Dan I can achieve the same result with Java

```
String content = 
"[source,bash]\n" +
"----\n" +
"$ curl 'http://localhost:8080/' -i\n" +
"----";

Document loadedDoc = asciidoctor.load(content, 
  OptionsBuilder.options().attributes(AttributesBuilder.attributes()
    attribute("safe", documentAttributes.get("safe"))
   .attribute("fragment", "")).asMap());

for (AbstractBlock abstractBlock : loadedDoc.getBlocks()) {
  parent.getBlocks().add(abstractBlock);
}
```

But there is still a thing that keeps me away from this: I keep getting an exception because the `SnippetsDirectoryResolver` cannot find the attribute `projectdir`. It is set on the document BUT even my single line unit test enters the resolver twice with another instance of a document where this attribute is missing?!

The second problem with the code above is that there is no way to set the safe mode but only options.
Maybe @mojavelinux has any additional hint for me.
For now I will start with a ruby solution for the problem.

**Comment by mojavelinux** [comment]
> I keep getting an exception because the SnippetsDirectoryResolver cannot find the attribute projectdir

When you call `Asciidoctor#load` inside the extension, you need to explicitly pass on any attributes you want from the original document. This new document instance doesn't have any visibility into the original document, so you have to think of it like loading a document in an isolated environment. But you can read the attributes from the original document and pass on the ones you want.

I'm looking into an API that would allow this information to be passed on automatically, but it needs some work still.

**Comment by mojavelinux** [comment]
> attribute("safe", documentAttributes.get("safe"))

safe is not an attribute. It's an option. (though I do think the value is exposed via the attribute `safe-mode-name`. But it has to be passed in as an option.

**Comment by mojavelinux** [comment]
> I have tested the approach above but it won't work with PDFs

It should work with any converter. That's the idea of populating the AST. Perhaps the extension isn't getting loaded correctly in the task configuration that uses the pdf backend.

**Comment by meistermeier** [comment]
Thanks @mojavelinux , somehow I was mixing up options and attributes.

For now I have the Ruby based extension running included in the spring-restdocs-asciidoctor module and used in one of the samples and it is including static input.
Next stop: loading snippets.

Btw. the same approach also works with Java but feels "not so right". Maybe I will show both ones in the end.

**Comment by meistermeier** [comment]
I really do not know if I did it the ruby way but there is a solution: https://github.com/meistermeier/spring-restdocs/blob/55eee65f1a41d7fb564cb48408f4998c9c662664/spring-restdocs-asciidoctor/src/main/resources/extensions/restdocs-block-macro.rb

Linking the diff again for an overview: https://github.com/spring-projects/spring-restdocs/compare/master...meistermeier:block_macro

The usage slightly changed and expects quotes around the snippets to get the correct attributes.
Just as a side note: I saw the right output in the pdf once. Will invest more time if you give me a go for the ruby code.
